### ISSUE TITLE
[Scheduler] handling failures

### ISSUE THREAD

**Issue by kbond** [symfony/symfony#49804]
### Symfony version(s) affected

6.3

### Description

When running `messenger:consume` for your schedule and a scheduled message handler throws an exception, there is a hard failure and the process exits. [This is the exception](https://github.com/symfony/symfony/blob/aadd302a0a2d2505979b9eb34e6767e98533b3ef/src/Symfony/Component/Scheduler/Messenger/SchedulerTransport.php#L43) that's thrown.

Is this expected? If not, what should we do?

### How to reproduce

Run `bin/console messenger:consume default_schedule` and have a handler throw an exception.

### Possible Solution

I'm thinking by default we should swallow the exception and move on? Of course dispatching relative events (#49803).

Perhaps this behaviour should be configurable?
- Retry
- Send to failures transport

**Comment by tucksaun** [comment]
I wanted to share my experience on this topic because I recently experienced the retry mechanism (because I configured the transport myself the retry mechanism became enabled automatically by Messenger)
In that case, one encounters another kind of failure: "cannot send messages" because this is not possible with the current SchedulerTransport. This raises the question of how retry would be possible.

But anyway, because `SchedulerTransport::reject` throws an exception, Messenger can not go on. If we remove the exception thrown in `reject`, Messenger is at least be able to report the initial error correctly and should be able to send the message to the failure transport if one is configured.

Regarding retries, I see three options here:
 - We could state that if retries are actually important for someone, then they should instead use a `RedispatchMessage` to dispatch the message to a transport that actually supports retries.
- We could let the user configure a retry transport where to send retries, and only forward messages with a scheduled and redeliveryStamp
- We could embed an in-memory transport to perform retries

Option 1 has the advantage of relying on all the usual Messenger infrastructure and keeping the Scheduler code simple.

Note that the last option is not safe as it does not persist errors across consumer reboots but it can be a fallback of option 2 and it looks like it works "okay" from a quick test I just did:
```
16:38:49 WARNING   [messenger] Error thrown while handling message App\Messenger\TestMessage. Sending for retry #1 using 1000 ms delay. Error: "Handling "App\Messenger\TestMessage" failed: " ["class" => "App\Messenger\TestMessage","retryCount" => 1,"delay" => 1000,"error" => "Handling "App\Messenger\TestMessage" failed: ","exception" => Symfony\Component\Messenger\Exception\HandlerFailedException^ { …}]
^ Symfony\Component\Messenger\Stamp\RedeliveryStamp^ {#1385
  -retryCount: 1
  -redeliveredAt: DateTimeImmutable @1680799129 {#1386
    date: 2023-04-06 16:38:49.555470 UTC (+00:00)
  }
}
16:38:50 WARNING   [messenger] Error thrown while handling message App\Messenger\TestMessage. Sending for retry #2 using 2000 ms delay. Error: "Handling "App\Messenger\TestMessage" failed: " ["class" => "App\Messenger\TestMessage","retryCount" => 2,"delay" => 2000,"error" => "Handling "App\Messenger\TestMessage" failed: ","exception" => Symfony\Component\Messenger\Exception\HandlerFailedException^ { …}]
^ Symfony\Component\Messenger\Stamp\RedeliveryStamp^ {#449
  -retryCount: 2
  -redeliveredAt: DateTimeImmutable @1680799130 {#407
    date: 2023-04-06 16:38:50.703626 UTC (+00:00)
  }
}

16:38:52 WARNING   [messenger] Error thrown while handling message App\Messenger\TestMessage. Sending for retry #3 using 4000 ms delay. Error: "Handling "App\Messenger\TestMessage" failed: " ["class" => "App\Messenger\TestMessage","retryCount" => 3,"delay" => 4000,"error" => "Handling "App\Messenger\TestMessage" failed: ","exception" => Symfony\Component\Messenger\Exception\HandlerFailedException^ { …}]
^ Symfony\Component\Messenger\Stamp\RedeliveryStamp^ {#955
  -retryCount: 3
  -redeliveredAt: DateTimeImmutable @1680799132 {#941
    date: 2023-04-06 16:38:52.816743 UTC (+00:00)
  }
}

16:38:56 CRITICAL  [messenger] Error thrown while handling message App\Messenger\TestMessage. Removing from transport after 3 retries. Error: "Handling "App\Messenger\TestMessage" failed: " ["class" => "App\Messenger\TestMessage","retryCount" => 3,"error" => "Handling "App\Messenger\TestMessage" failed: ","exception" => Symfony\Component\Messenger\Exception\HandlerFailedException^ { …}]
```

I have drafted some code locally so I can open a PR with this approach if you want

**Comment by kbond** [comment]
Thanks for your input here!

> If we remove the exception thrown in reject, Messenger is at least be able to report the initial error correctly and should be able to send the message to the failure transport if one is configured.

I feel like this is the easiest place to start. For my projects, this would be exactly what I want.

**Comment by tucksaun** [comment]
Yeah same here, I will open the PR later today for this specific bit then 👍

**Comment by tucksaun** [comment]
@kbond do you think we should close this issue now that #49964 is merged?

**Comment by kbond** [comment]
Closing as fixed by #49964.
